### PR TITLE
feat:[CDS-73315]: github app authentication connector changes

### PR DESCRIPTION
**Pull request by sourabh-awashti**
## Describe your changes
These changes are corresponding to adding support for github app based authentication in Github connector.
Tests Done:
1. Github connector creation and authentication with github app from both manager and delegate side.
## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/50132)

### CODE REVIEW

**Review by raghavharness — APPROVED** [review]
Looks good, please once validate backward compatibility by keeping all the managers new and delegate old.

**Comment by sourabh-awashti** [comment]
trigger uts

**Comment by sourabh-awashti** [comment]
trigger uts

File: 440-connector-nextgen/src/main/java/io/harness/connector/heartbeat/ScmConnectorValidationParamsProvider.java [hunk 1]
```diff
@@ -29,10 +34,13 @@
 @Singleton
 public class ScmConnectorValidationParamsProvider implements ConnectorValidationParamsProvider {
   @Inject GitConfigAuthenticationInfoHelper gitConfigAuthenticationInfoHelper;
+  @Inject private AccountClient accountClient;
 
   @Override
   public ConnectorValidationParams getConnectorValidationParams(ConnectorInfoDTO connectorInfoDTO, String connectorName,
       String accountIdentifier, String orgIdentifier, String projectIdentifier) {
+    boolean githubAppAuthentication = getResponse(
```
Comment (yogesh-chauhan) [inline 440-connector-nextgen/src/main/java/io/harness/connector/heartbeat/ScmConnectorValidationParamsProvider.java]: should be final

File: 440-connector-nextgen/src/main/java/io/harness/connector/heartbeat/ScmConnectorValidationParamsProvider.java [hunk 2]
```diff
@@ -50,11 +58,16 @@ public ConnectorValidationParams getConnectorValidationParams(ConnectorInfoDTO c
       encryptedDataDetails.addAll(
           gitConfigAuthenticationInfoHelper.getApiAccessEncryptedDataDetail(scmConnector, ngAccess));
     }
+    if (githubAppAuthentication && GitAuthenticationDecryptionHelper.isGitHubAppAuthentication(scmConnector)) {
```
Comment (yogesh-chauhan) [inline 440-connector-nextgen/src/main/java/io/harness/connector/heartbeat/ScmConnectorValidationParamsProvider.java]: We should call FF only if its of github app

File: 440-connector-nextgen/src/main/java/io/harness/connector/entities/embedded/githubconnector/GithubApp.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+@Value
+@Builder
+@TypeAlias("io.harness.connector.entities.embedded.githubconnector.GithubApp")
+public class GithubApp implements GithubHttpAuth {
+  String installationId;
+  String applicationId;
+  String privateKeyRef;
+  String installationIdRef;
+  String applicationIdRef;
```
Comment (yogesh-chauhan) [inline 440-connector-nextgen/src/main/java/io/harness/connector/entities/embedded/githubconnector/GithubApp.java:22]: lets annotate with not null

File: 440-connector-nextgen/src/main/java/io/harness/connector/validator/scmValidators/AbstractGitConnectorValidator.java [hunk 4]
```diff
@@ -72,7 +72,7 @@ public <T extends ConnectorConfigDTO> TaskParameters getTaskParameters(
         .build();
   }
 
-  private NGAccess getNgAccess(String accountIdentifier, String orgIdentifier, String projectIdentifier) {
+  public NGAccess getNgAccess(String accountIdentifier, String orgIdentifier, String projectIdentifier) {
```
Comment (yogesh-chauhan) [inline 440-connector-nextgen/src/main/java/io/harness/connector/validator/scmValidators/AbstractGitConnectorValidator.java:75]: protected

File: 440-connector-nextgen/src/main/java/io/harness/connector/validator/scmValidators/GithubConnectorValidator.java [hunk 5]
```diff
@@ -7,19 +7,52 @@
...
+  @Inject GitConfigAuthenticationInfoHelper gitConfigAuthenticationInfoHelper;
+
   @Override
   public GitConfigDTO getGitConfigFromConnectorConfig(ConnectorConfigDTO connectorConfig) {
     return GithubToGitMapper.mapToGitConfigDTO((GithubConnectorDTO) connectorConfig);
   }
 
+  @Override
+  public <T extends ConnectorConfigDTO> TaskParameters getTaskParameters(
+      T connectorConfig, String accountIdentifier, String orgIdentifier, String projectIdentifier) {
```
Comment (yogesh-chauhan) [inline 440-connector-nextgen/src/main/java/io/harness/connector/validator/scmValidators/GithubConnectorValidator.java:39]: <T extends ConnectorConfigDTO> -> can this be GithubConnectorDTO

File: 440-connector-nextgen/src/main/java/io/harness/connector/validator/scmValidators/GithubConnectorValidator.java [hunk 6]
```diff
@@ -7,19 +7,52 @@
...
+    List<EncryptedDataDetail> authenticationEncryptedDataDetails = gitCommandParams.getEncryptionDetails();
+    GithubConnectorDTO githubConnectorDTO = (GithubConnectorDTO) connectorConfig;
+    if (GitAuthenticationDecryptionHelper.isGitHubAppAuthentication(githubConnectorDTO)) {
+      authenticationEncryptedDataDetails.addAll(gitConfigAuthenticationInfoHelper.getGithubAppEncryptedDataDetail(
+          githubConnectorDTO, super.getNgAccess(accountIdentifier, orgIdentifier, projectIdentifier)));
+    }
+
+    gitCommandParams.setGithubAppAuthentication(getResponse(
+        accountClient.isFeatureFlagEnabled(FeatureName.CDS_GITHUB_APP_AUTHENTICATION.name(), accountIdentifier)));
+    gitCommandParams.setEncryptionDetails(authenticationEncryptedDataDetails);
```
Comment (yogesh-chauhan) [inline 440-connector-nextgen/src/main/java/io/harness/connector/validator/scmValidators/GithubConnectorValidator.java:55]: can this move inside the IF block

File: 440-connector-nextgen/src/main/java/io/harness/connector/entities/embedded/githubconnector/GithubApp.java (hunk 3 above)
Comment (sourabh-awashti) [reply 440-connector-nextgen/src/main/java/io/harness/connector/entities/embedded/githubconnector/GithubApp.java:23]: done

File: 440-connector-nextgen/src/main/java/io/harness/connector/validator/scmValidators/GithubConnectorValidator.java (hunk 5 above)
Comment (sourabh-awashti) [reply 440-connector-nextgen/src/main/java/io/harness/connector/validator/scmValidators/GithubConnectorValidator.java:39]: tried but giving error

File: 440-connector-nextgen/src/main/java/io/harness/connector/heartbeat/ScmConnectorValidationParamsProvider.java (hunk 2 above)
Comment (sourabh-awashti) [reply 440-connector-nextgen/src/main/java/io/harness/connector/heartbeat/ScmConnectorValidationParamsProvider.java]: done

File: 440-connector-nextgen/src/main/java/io/harness/connector/validator/scmValidators/AbstractGitConnectorValidator.java (hunk 4 above)
Comment (sourabh-awashti) [reply 440-connector-nextgen/src/main/java/io/harness/connector/validator/scmValidators/AbstractGitConnectorValidator.java]: done

File: 440-connector-nextgen/src/main/java/io/harness/connector/validator/scmValidators/GithubConnectorValidator.java (hunk 6 above)
Comment (sourabh-awashti) [reply 440-connector-nextgen/src/main/java/io/harness/connector/validator/scmValidators/GithubConnectorValidator.java]: done

**Comment by sourabh-awashti** [comment]
*[`440-connector-nextgen/src/main/java/io/harness/connector/heartbeat/ScmConnectorValidationParamsProvider.java` line 42 at r1](https://reviewable.io/reviews/harness/harness-core/50132#gh-1255358899:-NZkFlV30ile3-axUh6K:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/cdb46c206177d3240ebef8c8c3cde02f66b73ac7/440-connector-nextgen/src/main/java/io/harness/connector/heartbeat/ScmConnectorValidationParamsProvider.java#L42)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

should be final
</blockquote></details>

Done.

**Comment by sourabh-awashti** [comment]
trigger uts

File: 440-connector-nextgen/src/main/java/io/harness/connector/heartbeat/ScmConnectorValidationParamsProvider.java (hunk 1 above)
Comment (sourabh-awashti) [reply 440-connector-nextgen/src/main/java/io/harness/connector/heartbeat/ScmConnectorValidationParamsProvider.java]: done

**Review by yogesh-chauhan — DISMISSED** [review]
> Reviewed 19 of 30 files at r1, 11 of 11 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/50132)* status: 30 file(s) to review. 1 of 2 approvals obtained.

**Comment by sourabh-awashti** [comment]
trigger uts

**Comment by sourabh-awashti** [comment]
trigger codeformat

**Comment by sourabh-awashti** [comment]
trigger checkstylecodeformat

**Comment by sourabh-awashti** [comment]
trigger uts

**Comment by sourabh-awashti** [comment]
trigger ut0

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 4 of 4 files at r4, 4 of 4 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/50132)* status: 33 file(s) to review. 1 of 2 approvals obtained.

**Review by raghavharness — APPROVED** [review]

**Comment by sourabh-awashti** [comment]
trigger review

**Pull request merged**